File: app/src/main/java/com/slimshady/bottomsheetmodule/adapter/AsyncSearchListAdapter.java
package com.slimshady.bottomsheetmodule.adapter;


import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.AsyncListDiffer;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.slimshady.bottomsheetmodule.adapter.diffUtil.SearchListAdapterDiffUtil;
import com.slimshady.bottomsheetmodule.adapter.listener.AdapterItemListener;
import com.slimshady.bottomsheetmodule.databinding.ItemBaseSearchBinding;
import com.slimshady.bottomsheetmodule.model.BaseBottomSheetRecyclerModel;
import com.slimshady.bottomsheetmodule.util.AdapterAction;

import java.util.ArrayList;
import java.util.List;

public class AsyncSearchListAdapter<T extends BaseBottomSheetRecyclerModel> extends ListAdapter<T, AsyncSearchListAdapter<T>.ViewHolder> {

    private boolean isMultiSelect;
    private final AsyncListDiffer<T> mDiffer = new AsyncListDiffer<T>(this, new SearchListAdapterDiffUtil<T>());
    private AdapterItemListener<T> adapterItemListener;


    public AsyncSearchListAdapter(boolean isMultiSelect, AdapterItemListener<T> adapterItemListener) {

        super(new SearchListAdapterDiffUtil<T>());
        this.isMultiSelect = isMultiSelect;
        this.adapterItemListener = adapterItemListener;

    }


    public AsyncSearchListAdapter(boolean isMultiSelect) {

        super(new SearchListAdapterDiffUtil<T>());
        this.isMultiSelect = isMultiSelect;
    }

    @NonNull
    @Override
    public AsyncSearchListAdapter<T>.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ItemBaseSearchBinding itemBaseSearchBinding =
                ItemBaseSearchBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new AsyncSearchListAdapter<T>.ViewHolder(itemBaseSearchBinding);
    }

    @Override
    public void onBindViewHolder(@NonNull AsyncSearchListAdapter.ViewHolder holder, int position) {
        holder.setIsRecyclable(false);
        holder.bind(mDiffer.getCurrentList().get(position));

    }

    @Override
    public int getItemCount() {
        return mDiffer.getCurrentList().size();
    }

    @Override
    public void submitList(List<T> list) {
        mDiffer.submitList(list);
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        ItemBaseSearchBinding binding;

        public ViewHolder(@NonNull ItemBaseSearchBinding binding) {
            super(binding.getRoot());
            this.binding = binding;

        }

        public void bind(T baseSearchDbModel) {
            binding.txtSearch.setText(baseSearchDbModel.getName());

            if (!isMultiSelect) {
                binding.checkboxSearch.setVisibility(View.GONE);
            }

            binding.getRoot().setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (isMultiSelect)
                        toggleSelection(baseSearchDbModel, binding);
                    else
                        adapterItemListener.onItemSelect(baseSearchDbModel, getAbsoluteAdapterPosition(), AdapterAction.SELECT);
                }
            });

        }
    }

    boolean isSelected(int position) {
        return mDiffer.getCurrentList().get(position).isSelected();
    }

    public void toggleSelection(T baseSearchDbModel, ItemBaseSearchBinding binding) {
        if (!baseSearchDbModel.isSelected()) {
            baseSearchDbModel.setSelected(true);
            binding.checkboxSearch.setChecked(true);

        } else {
            baseSearchDbModel.setSelected(false);
            binding.checkboxSearch.setChecked(false);

        }

    }

    public void selectAll() {
        for (int i = 0; i < mDiffer.getCurrentList().size(); i++) {
            mDiffer.getCurrentList().get(i).setSelected(true);

        }
    }

    public void clearSelections() {
        for (int i = 0; i < mDiffer.getCurrentList().size(); i++) {
            mDiffer.getCurrentList().get(i).setSelected(false);

        }

    }

    public int getSelectedItemCount() {
        int count = 0;

        for (int i = 0; i < mDiffer.getCurrentList().size(); i++) {
            if (mDiffer.getCurrentList().get(i).isSelected())
                count++;

        }

        return count;
    }

    public ArrayList<T> getSelectedItems() {
        ArrayList<T> mDifferTemp = new ArrayList<>();

        for (int i = 0; i < mDiffer.getCurrentList().size(); i++) {
            if (mDiffer.getCurrentList().get(i).isSelected())
                mDifferTemp.add(mDiffer.getCurrentList().get(i));

        }
        return mDifferTemp;

    }
}

File: app/src/main/java/com/slimshady/bottomsheetmodule/adapter/AsyncSearchListAdapter.kt
package com.slimshady.bottomsheetmodule.adapter

import com.slimshady.bottomsheetmodule.model.BaseBottomSheetRecyclerModel
import androidx.recyclerview.widget.AsyncListDiffer
import com.slimshady.bottomsheetmodule.adapter.diffUtil.SearchListAdapterDiffUtil
import com.slimshady.bottomsheetmodule.adapter.listener.AdapterItemListener
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.slimshady.bottomsheetmodule.databinding.ItemBaseSearchBinding
import com.slimshady.bottomsheetmodule.util.AdapterAction
import java.util.ArrayList

class AsyncSearchListAdapter<T : BaseBottomSheetRecyclerModel> :
    ListAdapter<T, AsyncSearchListAdapter<T>.ViewHolder> {
    private var isMultiSelect: Boolean
    private val mDiffer = AsyncListDiffer(this, SearchListAdapterDiffUtil<T>())
    private var adapterItemListener: AdapterItemListener<T>? = null

    constructor(isMultiSelect: Boolean, adapterItemListener: AdapterItemListener<T>?) : super(
        SearchListAdapterDiffUtil<T>()
    ) {
        this.isMultiSelect = isMultiSelect
        this.adapterItemListener = adapterItemListener
    }

    constructor(isMultiSelect: Boolean) : super(SearchListAdapterDiffUtil<T>()) {
        this.isMultiSelect = isMultiSelect
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemBaseSearchBinding =
            ItemBaseSearchBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return ViewHolder(itemBaseSearchBinding)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setIsRecyclable(false)
        holder.bind(mDiffer.currentList[position])
    }

    override fun getItemCount(): Int {
        return mDiffer.currentList.size
    }

    override fun submitList(list: List<T>?) {
        mDiffer.submitList(list)
    }

    inner class ViewHolder(var binding: ItemBaseSearchBinding) : RecyclerView.ViewHolder(
        binding.root
    ) {
        fun bind(baseSearchDbModel: T) {
            binding.txtSearch.text = baseSearchDbModel.name
            if (!isMultiSelect) {
                binding.checkboxSearch.visibility = View.GONE
            }
            binding.root.setOnClickListener {
                if (isMultiSelect) toggleSelection(
                    baseSearchDbModel,
                    binding
                ) else adapterItemListener?.onItemSelect(
                    baseSearchDbModel,
                    absoluteAdapterPosition,
                    AdapterAction.SELECT
                )
            }
        }
    }

    fun isSelected(position: Int): Boolean {
        return mDiffer.currentList[position]?.isSelected ?: false
    }

    fun toggleSelection(baseSearchDbModel: T, binding: ItemBaseSearchBinding) {
        if (baseSearchDbModel.isSelected) {
            baseSearchDbModel.isSelected = false
            binding.checkboxSearch.isChecked = false

        } else {
            baseSearchDbModel.isSelected = true
            binding.checkboxSearch.isChecked = true

        }
    }

    fun selectAll() {
        for (i in mDiffer.currentList.indices) {
            mDiffer.currentList[i]?.isSelected = true
        }
    }

    fun clearSelections() {
        for (i in mDiffer.currentList.indices) {
            mDiffer.currentList[i]?.isSelected = false
        }
    }

    val selectedItemCount: Int
        get() {
            var count = 0
            for (i in mDiffer.currentList.indices) {
                if (mDiffer.currentList[i]?.isSelected == true) count++
            }
            return count
        }
    val selectedItems: ArrayList<T>
        get() {
            val mDifferTemp = ArrayList<T>()
            for (i in mDiffer.currentList.indices) {
                if (mDiffer.currentList[i]?.isSelected == true) mDifferTemp.add(mDiffer.currentList[i])
            }
            return mDifferTemp
        }
}